pkg/pillar/devicenetwork/devicenetwork.go

```diff
@@ -385,30 +385,52 @@ func devPortInstallWifiConfig(ifname string, wconfig types.WirelessConfig) bool
 
 func getWifiCredential(wifi types.WifiConfig) (zconfig.CredentialBlock, error) {
 	cred := zconfig.CredentialBlock{}
-	if !wifi.IsCipher {
-		cred.Identity = wifi.Identity
-		cred.Password = wifi.Password
-		return cred, nil
-	}
-	if !wifi.IsValidCipher {
-		errStr := fmt.Sprintf("%s, Cipher Block is not ready", wifi.SSID)
+	cipherBlock := new(types.CipherBlockStatus)
+	*cipherBlock = wifi.CipherBlockStatus
+	if !cipherBlock.IsCipher {
+		return handleCipherBlockError(cipherBlock, wifi.Identity,
+			wifi.Password, nil)
+	}
+	if len(wifi.Error) != 0 {
+		errStr := fmt.Sprintf("%s, Cipher Block is not ready, %s",
+			wifi.SSID, cipherBlock.Error)
 		log.Errorln(errStr)
-		return cred, errors.New(errStr)
+		err := errors.New(errStr)
+		return handleCipherBlockError(cipherBlock, wifi.Identity,
+			wifi.Password, err)
 	}
-	clearData, err := tpmmgr.DecryptCipherBlock(wifi.CipherBlock)
+	clearData, err := tpmmgr.DecryptCipherBlock(*cipherBlock)
 	if err != nil {
 		log.Errorf("%s, wifi CipherData decryption failed, %v\n",
 			wifi.SSID, err)
-		return cred, err
+		return handleCipherBlockError(cipherBlock, wifi.Identity,
+			wifi.Password, err)
 	}
-	if err := proto.Unmarshal(clearData, &cred); err != nil {
+	if err = proto.Unmarshal(clearData, &cred); err != nil {
 		log.Errorf("%s, wifi credential unmarshall failed, %v\n",
 			wifi.SSID, err)
-		return cred, err
+		return handleCipherBlockError(cipherBlock, wifi.Identity,
+			wifi.Password, err)
 	}
 	return cred, nil
 }
 
+func handleCipherBlockError(status *types.CipherBlockStatus,
```

Comment:
You have two copies of this. Can it go in the utils package?

Reply:
Or as a meber function of type.CipheblockStatus?

Reply:
Have moved it to utils (cipherutils.go)